### ISSUE TITLE
MinGW build broken

### ISSUE THREAD

**Issue by hgdagon** [ihhub/fheroes2#1646]
I tried building with msys2, and got these errors.

SDL2:

```Bash

audio_cdrom.cpp:62:57: error: macro "SDL_CreateThread" requires 3 arguments, but only 2 given
   62 |         loopThread = SDL_CreateThread( &LoopCheck, NULL );
      |                                                         ^
In file included from D:/msys64/mingw64/include/SDL2/SDL_audio.h:35,
                 from D:/msys64/mingw64/include/SDL2/SDL.h:36,
                 from types.h:26,
                 from audio_cdrom.h:27,
                 from audio_cdrom.cpp:25:
D:/msys64/mingw64/include/SDL2/SDL_thread.h:132: note: macro "SDL_CreateThread" defined here
  132 | #define SDL_CreateThread(fn, name, data) SDL_CreateThread(fn, name, data, (pfnSDL_CurrentBeginThread)SDL_beginthread, (pfnSDL_CurrentEndThread)SDL_endthread)
      |
audio_cdrom.cpp:34:12: error: 'SDL_CD' does not name a type; did you mean 'SDL_h_'?
   34 |     static SDL_CD * cd = NULL;
      |            ^~~~~~
      |            SDL_h_
audio_cdrom.cpp: In function 'void Cdrom::Open()':
audio_cdrom.cpp:46:26: error: 'SDL_CDNumDrives' was not declared in this scope
   46 |     for ( int i = 0; i < SDL_CDNumDrives(); i++ ) {
      |                          ^~~~~~~~~~~~~~~
audio_cdrom.cpp:47:9: error: 'SDL_CD' was not declared in this scope; did you mean 'SDL_h_'?
   47 |         SDL_CD * drive = SDL_CDOpen( i );
      |         ^~~~~~
      |         SDL_h_
audio_cdrom.cpp:47:18: error: 'drive' was not declared in this scope; did you mean 'div'?
   47 |         SDL_CD * drive = SDL_CDOpen( i );
      |                  ^~~~~
      |                  div
audio_cdrom.cpp:47:26: error: 'SDL_CDOpen' was not declared in this scope
   47 |         SDL_CD * drive = SDL_CDOpen( i );
      |                          ^~~~~~~~~~
audio_cdrom.cpp:50:31: error: 'SDL_CDStatus' was not declared in this scope; did you mean 'SDL_AudioStatus'?
   50 |             if ( !CD_INDRIVE( SDL_CDStatus( drive ) ) ) {
      |                               ^~~~~~~~~~~~
      |                               SDL_AudioStatus
audio_cdrom.cpp:50:19: error: 'CD_INDRIVE' was not declared in this scope
   50 |             if ( !CD_INDRIVE( SDL_CDStatus( drive ) ) ) {
      |                   ^~~~~~~~~~
audio_cdrom.cpp:51:17: error: 'SDL_CDClose' was not declared in this scope; did you mean 'SDL_RWclose'?
   51 |                 SDL_CDClose( drive );
      |                 ^~~~~~~~~~~
      |                 SDL_RWclose
audio_cdrom.cpp:54:71: error: 'SDL_DATA_TRACK' was not declared in this scope
   54 |             else if ( drive->numtracks > 1 && drive->track[0].type == SDL_DATA_TRACK ) {
      |                                                                       ^~~~~~~~~~~~~~
audio_cdrom.cpp:55:17: error: 'cd' was not declared in this scope
   55 |                 cd = drive;
      |                 ^~
audio_cdrom.cpp:61:10: error: 'cd' was not declared in this scope
   61 |     if ( cd ) {
      |          ^~
audio_cdrom.cpp:62:22: error: cannot convert 'SDL_Thread*(SDL_ThreadFunction, const char*, void*, pfnSDL_CurrentBeginThread, pfnSDL_CurrentEndThread)' {aka 'SDL_Thread*(int (*)(void*), const char*, void*, long long unsigned int (*)(void*, unsigned int, unsigned int (*)(void*), void*, unsigned int, unsigned int*), void (*)(unsigned int))'} to 'SDL_Thread*' in assignment
   62 |         loopThread = SDL_CreateThread( &LoopCheck, NULL );
      |                      ^~~~~~~~~~~~~~~~
      |                      |
      |                      SDL_Thread*(SDL_ThreadFunction, const char*, void*, pfnSDL_CurrentBeginThread, pfnSDL_CurrentEndThread) {aka SDL_Thread*(int (*)(void*), const char*, void*, long long unsigned int (*)(void*, unsigned int, unsigned int (*)(void*), void*, unsigned int, unsigned int*), void (*)(unsigned int))}
In file included from audio_cdrom.cpp:27:
audio_cdrom.cpp:66:12: error: 'cd' was not declared in this scope
   66 |     ERROR( cd ? "found CD audio device." : "no CDROM devices available." );
      |            ^~
system.h:49:22: note: in definition of macro 'COUT'
   49 |         std::cerr << x << std::endl;                                                                                                                                     \
      |                      ^
audio_cdrom.cpp:66:5: note: in expansion of macro 'ERROR'
   66 |     ERROR( cd ? "found CD audio device." : "no CDROM devices available." );
      |     ^~~~~
system.h:49:24: error: invalid operands of types 'const char [28]' and '<unresolved overloaded function type>' to binary 'operator<<'
   49 |         std::cerr << x << std::endl;                                                                                                                                     \
system.h:59:9: note: in expansion of macro 'COUT'
   59 |         COUT( System::GetTime() << ": [ERROR]\t" << __FUNCTION__ << ":  " << x );                                                                                        \
      |         ^~~~
audio_cdrom.cpp:66:5: note: in expansion of macro 'ERROR'
   66 |     ERROR( cd ? "found CD audio device." : "no CDROM devices available." );
      |     ^~~~~
audio_cdrom.cpp: In function 'void Cdrom::Close()':
audio_cdrom.cpp:71:10: error: 'cd' was not declared in this scope
   71 |     if ( cd ) {
      |          ^~
audio_cdrom.cpp:72:9: error: 'SDL_CDStop' was not declared in this scope
   72 |         SDL_CDStop( cd );
      |         ^~~~~~~~~~
audio_cdrom.cpp:73:9: error: 'SDL_KillThread' was not declared in this scope; did you mean 'SDL_WaitThread'?
   73 |         SDL_KillThread( loopThread );
      |         ^~~~~~~~~~~~~~
      |         SDL_WaitThread
audio_cdrom.cpp:75:9: error: 'SDL_CDClose' was not declared in this scope; did you mean 'SDL_RWclose'?
   75 |         SDL_CDClose( cd );
      |         ^~~~~~~~~~~
      |         SDL_RWclose
audio_cdrom.cpp: In function 'bool Cdrom::isValid()':
audio_cdrom.cpp:82:12: error: 'cd' was not declared in this scope
   82 |     return cd;
      |            ^~
audio_cdrom.cpp: In function 'int Cdrom::LoopCheck(void*)':
audio_cdrom.cpp:91:19: warning: conversion from 'int' to 'u8' {aka 'unsigned char'} may change value [-Wconversion]
   91 |             Play( currentTrack, true, true );
      |                   ^~~~~~~~~~~~
audio_cdrom.cpp:85:30: warning: unused parameter 'data' [-Wunused-parameter]
   85 | int Cdrom::LoopCheck( void * data )
      |                       ~~~~~~~^~~~
audio_cdrom.cpp: In function 'void Cdrom::Play(u8, bool, bool)':
audio_cdrom.cpp:99:30: error: 'cd' was not declared in this scope
   99 |     if ( Mixer::isValid() && cd ) {
      |                              ^~
audio_cdrom.cpp:103:18: error: 'SDL_CDPlayTracks' was not declared in this scope
  103 |             if ( SDL_CDPlayTracks( cd, track, 0, 1, 0 ) < 0 )
      |                  ^~~~~~~~~~~~~~~~
audio_cdrom.cpp:108:74: error: 'CD_FPS' was not declared in this scope
  108 |                 tickLength = (unsigned int)( ( cd->track[track].length / CD_FPS ) * 0.01f );
      |                                                                          ^~~~~~
audio_cdrom.cpp:114:18: error: 'SDL_CDStatus' was not declared in this scope; did you mean 'SDL_AudioStatus'?
  114 |             if ( SDL_CDStatus( cd ) != CD_PLAYING )
      |                  ^~~~~~~~~~~~
      |                  SDL_AudioStatus
audio_cdrom.cpp:114:40: error: 'CD_PLAYING' was not declared in this scope
  114 |             if ( SDL_CDStatus( cd ) != CD_PLAYING )
      |                                        ^~~~~~~~~~
audio_cdrom.cpp: In function 'void Cdrom::Pause()':
audio_cdrom.cpp:124:10: error: 'cd' was not declared in this scope
  124 |     if ( cd )
      |          ^~
audio_cdrom.cpp:125:9: error: 'SDL_CDPause' was not declared in this scope
  125 |         SDL_CDPause( cd );
      |         ^~~~~~~~~~~
make[2]: *** [Makefile:14: audio_cdrom.o] Error 1
make[2]: Leaving directory '/home/George/fheroes2/src/engine'
make[1]: *** [Makefile:111: all] Error 2
make[1]: Leaving directory '/home/George/fheroes2/src'
make: *** [Makefile:24: all] Error 2

```

SDL1:

```Bash

In file included from audio_cdrom.cpp:27:
audio_cdrom.cpp: In function 'void Cdrom::Open()':
system.h:49:24: error: invalid operands of types 'const char [28]' and '<unresolved overloaded function type>' to binary 'operator<<'
   49 |         std::cerr << x << std::endl;                                                                                                                                     \
system.h:59:9: note: in expansion of macro 'COUT'
   59 |         COUT( System::GetTime() << ": [ERROR]\t" << __FUNCTION__ << ":  " << x );                                                                                        \
      |         ^~~~
audio_cdrom.cpp:66:5: note: in expansion of macro 'ERROR'
   66 |     ERROR( cd ? "found CD audio device." : "no CDROM devices available." );
      |     ^~~~~
audio_cdrom.cpp: In function 'int Cdrom::LoopCheck(void*)':
audio_cdrom.cpp:91:19: warning: conversion from 'int' to 'u8' {aka 'unsigned char'} may change value [-Wconversion]
   91 |             Play( currentTrack, true, true );
      |                   ^~~~~~~~~~~~
audio_cdrom.cpp:85:30: warning: unused parameter 'data' [-Wunused-parameter]
   85 | int Cdrom::LoopCheck( void * data )
      |                       ~~~~~~~^~~~
audio_cdrom.cpp: In function 'void Cdrom::Play(u8, bool, bool)':
audio_cdrom.cpp:108:72: warning: conversion from 'Uint32' {aka 'unsigned int'} to 'float' may change value [-Wconversion]
  108 |                 tickLength = (unsigned int)( ( cd->track[track].length / CD_FPS ) * 0.01f );
      |                                              ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
make[2]: *** [Makefile:14: audio_cdrom.o] Error 1
make[2]: Leaving directory '/home/George/fheroes2/src/engine'
make[1]: *** [Makefile:111: all] Error 2
make[1]: Leaving directory '/home/George/fheroes2/src'
make: *** [Makefile:24: all] Error 2


```

Since only Visual Studio is mentioned (and provided scripts for), I'm assuming, this is just for MinGW.

**Comment by ihhub** [comment]
Hi @hgdagon , we dropped MinGW support in favor of Visual Studio long time ago as this is not a common platform for development.

**Comment by hgdagon** [comment]
That's too bad. So even if I would fix this, you would not accept a PR for MinGW?

**Comment by ihhub** [comment]
Hi @hgdagon , we will accept it if it doesn't break anything else :)

**Comment by idshibanov** [comment]
Judging by the errors it tries to compile dead code triggered only if AUDIO_CD definition is set. Moreover that CD library is SDL1 only.

**Comment by hgdagon** [comment]
OK, so I tried building without the Audio CD support. The first change I made was getting rid of MinGW checks [here](https://github.com/ihhub/fheroes2/blob/620050d725df3dcd140a3ab35a05f09622291b5b/src/engine/tools.cpp#L811), it complains about `Char` to `Const Char` conversion. Another minor issue was that `WITH_ICONS` is looking for an ico file, which isn't there. After providing the ico file, the res is generated successfully.

Now, I'm not sure I completely understand what I'm doing, since I rely a little too much on CMake, to ever bother with makefiles, but, after a lot of Googling,  I added a lot of libraries to `Makefile.mingw` to get rid of undefined references, but I still didn't manage to build with either SDL.

Here's the SDL1 error:

```
g++  agg.o bin_info.o icn.o m82.o mus.o sprite.o til.o xmi.o ai_base.o ai_common.o ai_hero_action.o army.o army_bar.o army_troop.o battle_action.o battle_animation.o battle_arena.o battle_army.o battle_board.o battle_bridge.o battle_catapult.o battle_cell.o battle_command.o battle_dialogs.o battle_grave.o battle_interface.o battle_main.o battle_only.o battle_tower.o battle_troop.o buildinginfo.o captain.o castle.o castle_building.o castle_dialog.o castle_mageguild.o castle_tavern.o castle_town.o castle_well.o mageguild.o dialog_adventure.o dialog_arena.o dialog_armyinfo.o dialog_box.o dialog_buyboat.o dialog_chest.o dialog_file.o dialog_frameborder.o dialog_gameinfo.o dialog_giftresources.o dialog_guardian.o dialog_levelup.o dialog_marketplace.o dialog_message.o dialog_quickinfo.o dialog_recrut.o dialog_resolution.o dialog_resourceinfo.o dialog_selectcount.o dialog_selectfile.o dialog_selectitems.o dialog_selectscenario.o dialog_settings.o dialog_skillinfo.o dialog_spellinfo.o dialog_spriteinfo.o dialog_system.o dialog_thievesguild.o difficulty.o fheroes2.o game.o game_delays.o game_highscores.o game_hotkeys.o game_interface.o game_io.o game_loadgame.o game_mainmenu.o game_newgame.o game_over.o game_scenarioinfo.o game_startgame.o game_static.o game_video.o cursor.o interface_border.o interface_buttons.o interface_cpanel.o interface_events.o interface_focus.o interface_gamearea.o interface_icons.o interface_radar.o interface_status.o player_info.o skill_bar.o splitter.o statusbar.o text.o ui_button.o ui_tool.o direction.o heroes.o heroes_action.o heroes_base.o heroes_dialog.o heroes_indicator.o heroes_meeting.o heroes_move.o heroes_recruits.o heroes_spell.o route.o skill.o color.o kingdom.o kingdom_overview.o luck.o morale.o payment.o profit.o puzzle.o race.o speed.o week.o ground.o maps.o maps_actions.o maps_fileinfo.o maps_objects.o maps_tiles.o maps_tiles_quantity.o mp2.o position.o visit.o monster.o mounts.o objcrck.o objdirt.o objdsrt.o objgras.o objlava.o objmult.o objsnow.o objswmp.o objtown.o objwatr.o objxloc.o trees.o pocketpc_dialog.o artifact.o artifact_ultimate.o maps_text.o resource.o spell.o spell_book.o spell_storage.o gamedefs.o players.o settings.o test.o test_monstersprite.o world.o world_loadmap.o world_pathfinding.o ai_simple.o ai_simple_battle.o ai_simple_castle.o ai_simple_heroes.o ai_simple_kingdom.o ai_normal.o ai_normal_battle.o ai_normal_castle.o ai_normal_hero.o ai_normal_kingdom.o ../engine/libengine.a ../thirdparty/libsmacker/libsmacker.a ../xmlccwrap/libxmlccwrap.a fheroes2.res -static -Wl,-Bstatic -L/mingw64/lib -lmingw32 -lSDLmain -lSDL -mwindows -lSDL_mixer -lSDL_image -lpng16 -lz -lSDL_ttf  -lz -lSDL -lmad -lvorbisfile -lvorbis -logg -lSDL -lpng -lwebp -ltiff -ljpeg  -lfreetype -liconv -lharfbuzz -lgraphite2 -lfontconfig -llzma -lbrotlidec-static -lbrotlienc-static  -lbrotlicommon-static -lbz2 -lzstd -lwinmm -ldxguid -o fheroes2
D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/msys64/mingw64/lib\libharfbuzz.a(hb-ft.cc.obj):(.text+0x7f0): undefined reference to `FT_Get_MM_Var'
D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/msys64/mingw64/lib\libharfbuzz.a(hb-ft.cc.obj):(.text+0x84f): undefined reference to `FT_Get_Var_Blend_Coordinates'
D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/msys64/mingw64/lib\libharfbuzz.a(hb-ft.cc.obj):(.text+0x8c1): undefined reference to `FT_Done_MM_Var'
D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/msys64/mingw64/lib\libharfbuzz.a(hb-ft.cc.obj):(.text+0x1244): undefined reference to `FT_Set_Var_Blend_Coordinates'
D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/msys64/mingw64/lib\libharfbuzz.a(hb-uniscribe.cc.obj):(.text+0x706): undefined reference to `__imp_UuidCreate'
collect2.exe: error: ld returned 1 exit status
make[2]: *** [Makefile:46: fheroes2] Error 1
make[2]: Leaving directory '/home/George/fheroes2/src/dist'
make[1]: *** [Makefile:112: all] Error 2
make[1]: Leaving directory '/home/George/fheroes2/src'
make: *** [Makefile:24: all] Error 2
```

And here's SDL2:
```

g++  agg.o bin_info.o icn.o m82.o mus.o sprite.o til.o xmi.o ai_base.o ai_common.o ai_hero_action.o army.o army_bar.o army_troop.o battle_action.o battle_animation.o battle_arena.o battle_army.o battle_board.o battle_bridge.o battle_catapult.o battle_cell.o battle_command.o battle_dialogs.o battle_grave.o battle_interface.o battle_main.o battle_only.o battle_tower.o battle_troop.o buildinginfo.o captain.o castle.o castle_building.o castle_dialog.o castle_mageguild.o castle_tavern.o castle_town.o castle_well.o mageguild.o dialog_adventure.o dialog_arena.o dialog_armyinfo.o dialog_box.o dialog_buyboat.o dialog_chest.o dialog_file.o dialog_frameborder.o dialog_gameinfo.o dialog_giftresources.o dialog_guardian.o dialog_levelup.o dialog_marketplace.o dialog_message.o dialog_quickinfo.o dialog_recrut.o dialog_resolution.o dialog_resourceinfo.o dialog_selectcount.o dialog_selectfile.o dialog_selectitems.o dialog_selectscenario.o dialog_settings.o dialog_skillinfo.o dialog_spellinfo.o dialog_spriteinfo.o dialog_system.o dialog_thievesguild.o difficulty.o fheroes2.o game.o game_delays.o game_highscores.o game_hotkeys.o game_interface.o game_io.o game_loadgame.o game_mainmenu.o game_newgame.o game_over.o game_scenarioinfo.o game_startgame.o game_static.o game_video.o cursor.o interface_border.o interface_buttons.o interface_cpanel.o interface_events.o interface_focus.o interface_gamearea.o interface_icons.o interface_radar.o interface_status.o player_info.o skill_bar.o splitter.o statusbar.o text.o ui_button.o ui_tool.o direction.o heroes.o heroes_action.o heroes_base.o heroes_dialog.o heroes_indicator.o heroes_meeting.o heroes_move.o heroes_recruits.o heroes_spell.o route.o skill.o color.o kingdom.o kingdom_overview.o luck.o morale.o payment.o profit.o puzzle.o race.o speed.o week.o ground.o maps.o maps_actions.o maps_fileinfo.o maps_objects.o maps_tiles.o maps_tiles_quantity.o mp2.o position.o visit.o monster.o mounts.o objcrck.o objdirt.o objdsrt.o objgras.o objlava.o objmult.o objsnow.o objswmp.o objtown.o objwatr.o objxloc.o trees.o pocketpc_dialog.o artifact.o artifact_ultimate.o maps_text.o resource.o spell.o spell_book.o spell_storage.o gamedefs.o players.o settings.o test.o test_monstersprite.o world.o world_loadmap.o world_pathfinding.o ai_simple.o ai_simple_battle.o ai_simple_castle.o ai_simple_heroes.o ai_simple_kingdom.o ai_normal.o ai_normal_battle.o ai_normal_castle.o ai_normal_hero.o ai_normal_kingdom.o ../engine/libengine.a ../thirdparty/libsmacker/libsmacker.a ../xmlccwrap/libxmlccwrap.a fheroes2.res -static -Wl,-Bstatic -L/mingw64/lib -lmingw32 -lSDL2main -lSDL2 -mwindows -lSDL2_mixer -lSDL2_image -lpng16 -lz -lSDL2_ttf  -lz -lsetupapi -lversion -limm32 -lole32 -loleaut32 -lSDL2 -lmad -lvorbisfile -lvorbis -logg -lflac -lmpg123  -lshlwapi -lSDL2 -lpng -lwebp -ltiff -ljpeg  -lfreetype -liconv -lharfbuzz -lgraphite2 -lfontconfig -llzma -lbrotlidec-static -lbrotlienc-static  -lbrotlicommon-static -lbz2 -lzstd -lwinmm -ldxguid -o fheroes2
D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/msys64/mingw64/lib\libSDL2_mixer.a(music_opus.o):(.rdata$.refptr.op_pcm_seek[.refptr.op_pcm_seek]+0x0): undefined reference to `op_pcm_seek'
D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/msys64/mingw64/lib\libSDL2_mixer.a(music_opus.o):(.rdata$.refptr.op_read[.refptr.op_read]+0x0): undefined reference to `op_read'
D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/msys64/mingw64/lib\libSDL2_mixer.a(music_opus.o):(.rdata$.refptr.op_seekable[.refptr.op_seekable]+0x0): undefined reference to `op_seekable'
D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/msys64/mingw64/lib\libSDL2_mixer.a(music_opus.o):(.rdata$.refptr.op_head[.refptr.op_head]+0x0): undefined reference to `op_head'
D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/msys64/mingw64/lib\libSDL2_mixer.a(music_opus.o):(.rdata$.refptr.op_free[.refptr.op_free]+0x0): undefined reference to `op_free'
D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/msys64/mingw64/lib\libSDL2_mixer.a(music_opus.o):(.rdata$.refptr.op_open_callbacks[.refptr.op_open_callbacks]+0x0): undefined reference to `op_open_callbacks'
D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/msys64/mingw64/lib\libharfbuzz.a(hb-ft.cc.obj):(.text+0x7f0): undefined reference to `FT_Get_MM_Var'
D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/msys64/mingw64/lib\libharfbuzz.a(hb-ft.cc.obj):(.text+0x84f): undefined reference to `FT_Get_Var_Blend_Coordinates'
D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/msys64/mingw64/lib\libharfbuzz.a(hb-ft.cc.obj):(.text+0x8c1): undefined reference to `FT_Done_MM_Var'
D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/msys64/mingw64/lib\libharfbuzz.a(hb-ft.cc.obj):(.text+0x1244): undefined reference to `FT_Set_Var_Blend_Coordinates'
D:/msys64/mingw64/lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/msys64/mingw64/lib\libharfbuzz.a(hb-uniscribe.cc.obj):(.text+0x706): undefined reference to `__imp_UuidCreate'
collect2.exe: error: ld returned 1 exit status
make[2]: *** [Makefile:46: fheroes2] Error 1
make[2]: Leaving directory '/home/George/fheroes2/src/dist'
make[1]: *** [Makefile:112: all] Error 2
make[1]: Leaving directory '/home/George/fheroes2/src'
make: *** [Makefile:24: all] Error 2

```
And here's my [Makefile.mingw mess](https://github.com/ihhub/fheroes2/files/5214515/Makefile.zip).

.

**Comment by hgdagon** [comment]
OK, better results this time. I had massively underestimated how outdated `Makefile.mingw` was, so this time I just started from scratch instead of figuring out the existig choices. While I did achieve a completely static build, it required editing the main `Makefile`, while I'm trying to only edit `Makefile.mingw` to avoid potential issues with other platform, since I can't test all of them. However, if you're interested, here's some size comparison:

SDL1: 13101241 bytes/12.4MB
SDL1 Stripped: 9276928 bytes/8.84MB
SDL1 Stripped+UPX9: 4241408 bytes/4.04MB
SDL1 Dynamic: 6,407,833 bytes/6.11MB (+ dlls)
SDL2: 15185507 bytes/14.4MB
SDL2 Stripped: 11086336 bytes/10.5MB
SDL2 Stripped+UPX9: 4994048 bytes/4.76MB
SDL2 Dynamic: 6,405,649 bytes/6.10MB (+ dlls)

Instead, I opted for linking just the exotic mingw-specific libraries, as well as `iconv`, so the only dlls needed are `SDL`, `ogg`, `vorbis`, and `zlib`, which is already linked by the time `Makefile.mingw` is included.
Audio CD still doesn't work, along with network, but as far as I understand, they don't work for Linux (and MSVC?) either, and I was aiming to be as close to the Linux build, as posible. Also, tools are a bit of a mystery, I don't quite understand why they're all being linked against SDL. I had linking issues with `bin2txt`, and, even though I did eventually manage to build them, there's also an issue with the output. Again, doesn't seem like they're being used, so I just skipped those.

On a slightly unrelated note, the editor builds just fine, although, it could probably use a deploy script for windows dlls.

I'll make a PR with changes to `Makefile.mingw` soon, I just need one clarification, though. @ihhub @idshibanov Can I upload an icon? The build files reference it, but it doesn't exist. I chacked the old project over on SF, but the latest SVN still doesn't have it. I used an icon from the old SF build, and it seems to be the icon from the windows version of H2, which, I guess is why it was removed(?). I couldn't find a clear artwork of the H2 shield to even make a custom one, so I'm not sure what to do.

Finally here's some test builds, if anyone's interested. These are 64-bit [sdl1](https://github.com/ihhub/fheroes2/files/5233990/sdl1.zip) and [sdl2](https://github.com/ihhub/fheroes2/files/5233992/sdl2.zip).

**Comment by ihhub** [comment]
Hi @hgdagon , the source code contains it's own icon. All resources belong to the original game or other non-free stuff should not be a part of time project🙂

Also forget about tools build. This is very outdated code which we're planning to remove.

**Comment by hgdagon** [comment]
> Hi @hgdagon , the source code contains it's own icon.
@ihhub um, where, if I might ask?

**Comment by ihhub** [comment]
`fheroes2/image/embedded_image.h`

**Comment by hgdagon** [comment]
> `fheroes2/image/embedded_image.h`

@ihhub That's for the window. And, apparently, I made a mistake, the icon was not removed from the SVN, it was removed last week with [this](https://github.com/ihhub/fheroes2/commit/55d559df6db22c5ee50d7ce4d756f434fd04963f) commit, for licensing issues, like I thought. This means, the executables are left without an icon.
So, since the icon was removed deliberately, I will comment out icon related stuff as well, so that it can be enabled later when/if we get a custom icon.

**Comment by ihhub** [comment]
Hi @hgdagon , as I wrote previously the icon is in the above file. We generate and display it. It works perfectly on Windows under Visual Studio.

**Comment by hgdagon** [comment]
> It works perfectly on Windows under Visual Studio compilation.

@ihhub The executables in latest release (0.8.1) don't have any. The window does, but not the executables.

**Comment by ihhub** [comment]
Hi @hgdagon , is this issue still relevant?

**Comment by hgdagon** [comment]
@ihhub I'm not really sure, since I haven't been keeping up with FH2. The main issue here was building with Audio CD support. Does the code currently support CD audio? If it does, I should check again, otherwise, it's irrelevant.

Also, last I checked the CMake scripts weren't MSYS2/MinGW-friendly, I wanted to make a PR for that as well, but I haven't had the time yet. Makefile builds just fine for both 32 and 64 bit last I checked (without Audio CD support).
